### PR TITLE
Added JSON support for reading rules and selectors from pipeline

### LINKED ISSUE #857: Add JSON support for reading rules and selectors from pipeline
Extending from #833, we can add JSON support for reading rules and selectors from pipeline. Currently they can be exported to JSON, but not read fron JSON. 

We will need to create a an equivalent `LanguageExpression` deserializer to handle parsing `if` Selector expressions, which should be similar to the YAML implementation: https://github.com/microsoft/PSRule/blob/main/src/PSRule/Common/YamlConverters.cs#L611.

### PR DESCRIPTION
**Pull request by ArmaanMcleod**
## PR Summary

Fix #857

Added JSON support for reading rules and selectors. 

Also did some refactoring and cleanup with some previous baseline code. 

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Code changes**
  - [x] Have unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Microsoft/PSRule/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: src/PSRule/Common/JsonConverters.cs [hunk 1]
```diff
@@ -400,70 +401,75 @@ private IResource MapResource(JsonReader reader, JsonSerializer serializer)
...
+                        reader.Skip();
+                    }
+                }
+
+                else if (reader.TokenType == JsonToken.Comment)
+                {
+                    var commentLine = reader.Value.ToString().TrimStart();
 
-                        if (commentLine.StartsWith("Synopsis: "))
+                    if (commentLine.StartsWith("Synopsis: "))
```
Comment (BernieWhite) [inline src/PSRule/Common/JsonConverters.cs:463]: Maybe add a length check here as well.

```suggestion
                    if (commentLine.Length > 10 && commentLine.StartsWith("Synopsis: "))
```

**Review by BernieWhite — CHANGES_REQUESTED** [review]
@ArmaanMcleod Looking good. A few minor changes suggested. Otherwise all good.

I did notice there was a problem with synopsis reporting but this was an existing problem that affected YAML resources as well. I added a commit to resolve this. Please double check.

**Comment by ArmaanMcleod** [comment]
> @ArmaanMcleod Looking good. A few minor changes suggested. Otherwise all good.
> 
> I did notice there was a problem with synopsis reporting but this was an existing problem that affected YAML resources as well. I added a commit to resolve this. Please double check.

@BernieWhite Thanks for the feedback and also fixing the synopsis issue, love the new changes 🙂. 

Also fixed te docs a bit to emphasis using `.jsonc`. Let me know if you like these changes or would like to tweak it more.

File: docs/concepts/PSRule/en-US/about_PSRule_Selectors.md [hunk 2]
```diff
@@ -137,6 +155,67 @@ spec:
     - 'Value2'
 ```
 
+### Example Selectors.Rule.json
```
Comment (BernieWhite) [inline docs/concepts/PSRule/en-US/about_PSRule_Selectors.md:158]:
```suggestion
### Example Selectors.Rule.jsonc
```

File: docs/concepts/PSRule/en-US/about_PSRule_Selectors.md [hunk 3]
```diff
@@ -137,6 +155,67 @@ spec:
...
+
+```json
+// Example Selectors.Rule.json
```
Comment (BernieWhite) [inline docs/concepts/PSRule/en-US/about_PSRule_Selectors.md]:
```suggestion
// Example Selectors.Rule.jsonc
```

File: docs/concepts/PSRule/en-US/about_PSRule_Selectors.md [hunk 4]
```diff
@@ -73,10 +73,12 @@ If one or more selector pre-conditions are used, they are evaluated before type
 
 ### Defining selectors
 
-Selectors are defined in YAML and can be included within a module or standalone `.Rule.yaml` file.
-In either case, define a selector within a file ending with the `.Rule.yaml` extension.
+Selectors can be defined with either YAML or JSON format, and can be included with a module or standalone `.Rule.yaml` or `.Rule.jsonc` file.
+In either case, define a selector within a file ending with the `.Rule.yaml` or `.Rule.jsonc` extension.
 A selector can be defined side-by-side with other resources such as baselines or module configurations.
 
+Use `.jsonc` to view [JSON with Comments](https://code.visualstudio.com/docs/languages/json#_json-with-comments) in Visual Studio Code.
```
Comment (BernieWhite) [inline docs/concepts/PSRule/en-US/about_PSRule_Selectors.md:80]:
```suggestion
Selectors can also be defined within `.json` files.
We recommend using `.jsonc` to view [JSON with Comments](https://code.visualstudio.com/docs/languages/json#_json-with-comments) in Visual Studio Code.
```

File: docs/concepts/PSRule/en-US/about_PSRule_Rules.md [hunk 5]
```diff
@@ -156,6 +180,65 @@ spec:
...
 
+### Defining JSON rules
+
+To define a JSON rule use the `Rule` resource in a JSON file.
+Each rule must be defined within a `.Rule.jsonc` file following a standard schema.
+One or more rules can be defined in a single JSON array separating each rule in a JSON object.
+
+Within the `Rule` resource, the `condition` property specifies conditions to pass or fail the rule.
+
+Use the `.jsonc` extension to view [JSON with Comments](https://code.visualstudio.com/docs/languages/json#_json-with-comments) in Visual Studio Code.
```
Comment (BernieWhite) [inline docs/concepts/PSRule/en-US/about_PSRule_Rules.md:191]:
```suggestion
Rules can also be defined within `.json` files.
We recommend using `.jsonc` to view [JSON with Comments](https://code.visualstudio.com/docs/languages/json#_json-with-comments) in Visual Studio Code.
```

**Review by BernieWhite — APPROVED** [review]
@ArmaanMcleod Nice! A few minor suggestions. Otherwise all good to merge.

**Pull request merged**